Tools/clinic/clinic.py

```diff
@@ -540,13 +552,17 @@ def permute_right_option_groups(l):
       (1, 2, 3)
     """
     yield tuple()
-    accumulator = []
+    accumulator: list[Parameter] = []
     for group in l:
         accumulator.extend(group)
         yield tuple(accumulator)
 
 
-def permute_optional_groups(left, required, right):
+def permute_optional_groups(
+        left: Sequence[ParamTuple],
+        required: ParamTuple,
+        right: Sequence[ParamTuple]
```

Comment:
Mandating that the first and third parameters have to be passed `Sequence[tuple[Parameter, ...]]` rather than `Sequence[Iterable[Parameter]]` makes the type hints simpler now, but it will cause problems later on! You'll notice that this function is only ever actually _called_ once in `clinic.py`, and the one time it _is_ called... it's passed a `list[list[Parameter]]` for the first and third parameter!

https://github.com/python/cpython/blob/eeb5c63179eba0390c8937d0803a41092225be7e/Tools/clinic/clinic.py#L1371

So, you'll either need to change the annotations here, or change the place where the function is called so that it's passed objects of `list[tuple[Parameter, ...]]` instead of objects of type `list[list[Parameter]]`

Reply:
(The reason why mypy isn't flagging this issue now is because we have settings enabled so that mypy will only check functions with type annotations, and the place where this function is called is in a function that doesn't have type annotations.)

Reply:
Darn, you're right. There is no easy way out of this.

Reply:
Yeah -- this is why I used different types for the return types and parameter types in the diff I suggested in https://github.com/python/cpython/pull/106431#issuecomment-1620696237 :)

Reply:
Ok, PTAL! I broke this up in two commits; I'm not sure you'll like the latter (8d1821c26aa29e4bc9a1a83bb853d83ba17e9920) :)

Reply:
Looks great! Hard to find perfect names for these type aliases, but I think the aliases themselves do make sense; I'm fine with 8d1821c :)

Just one nit: https://github.com/python/cpython/pull/106431#discussion_r1252375853